casper/src/main/scala/coop/rchain/casper/engine/Running.scala

```diff
@@ -201,11 +202,14 @@ object Running {
         )
     )
 
-  def handleBlockMessage[F[_]: Monad: Log: RequestedBlocks](peer: PeerNode, b: BlockMessage)(
+  def handleBlockMessage[F[_]: Monad: Log: RequestedBlocks: BlockStore](
+      peer: PeerNode,
+      b: BlockMessage
+  )(
       repeatedCasperMessage: BlockHash => F[Boolean],
       casperAdd: BlockMessage => F[ValidBlockProcessing]
   ): F[Unit] =
-    repeatedCasperMessage(b.blockHash)
+    (repeatedCasperMessage(b.blockHash) ||^ BlockStore[F].contains(b.blockHash))
```

Comment:
Might be this should be part of `repeatedCasperMessage` itself

Reply:
Makes sense.

Reply:
Actually no - repeatedCasperMessage means casper message is being processed and node should not process it again. If node is restarted but block is in blockstore - node should process it again if it is still not in casper.